### PR TITLE
Port examples to sphinx_gallery

### LINKED ISSUE #167: Move notebooks to sphinx gallery python scripts
Much better way to have example notebooks compatible with a version control system like git
https://sphinx-gallery.github.io/stable/syntax.html

### PR DESCRIPTION
**Pull request by PicoCentauri**
Fixes #167 

As already stated in the issue this will allow easier maintance and gives these nice gallery view:

Screenshot 2023-03-25 at 10 30 47

Also i fixed some warnings when generating the docs.

### CODE REVIEW

**Comment by PicoCentauri** [comment]
Somebody can help me with the linter? I ran black locally but still got an error in the ci...

**Comment by PicoCentauri** [comment]
I just saw that we can even use [jupyterlite](https://sphinx-gallery.github.io/stable/configuration.html#jupyterlite) allowing running the examples in the browser. Should we try this?

File: examples/PlotLFRE.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import numpy as np
+
+from skmatter.datasets import load_degenerate_CH4_manifold
+from skmatter.metrics import pointwise_local_reconstruction_error
+
+mpl.rc("font", size=20)
+
+# %%
+#
+# load features
```
Comment (rosecers) [inline examples/PlotLFRE.py:35]: Same issue as last example

File: examples/PlotPointwiseGFRE.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    global_reconstruction_error,
+    pointwise_global_reconstruction_error,
+)
+from skmatter.preprocessing import StandardFlexibleScaler
+
+mpl.rc("font", size=20)
+
+# %%
+#
+# load features
```
Comment (rosecers) [inline examples/PlotPointwiseGFRE.py:38]: same as last two

File: examples/PlotPointwiseGFRE.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+#
+# We compare 3-body features with their mapping to the reproducing kernel Hilbert space
+# (RKHS) projected to the sample space using the nonlinear radial basis function (RBF) kernel
+#
+# .. math::
+#   k^{\textrm{RBF}}(\mathbf{x},\mathbf{x}') =
+#       \exp(-\gamma \|\mathbf{x}-\mathbf{x}'\|^2),\quad \gamma\in\mathbb{R}_+
+#
+# The projected RKHS features are computed using the eigendecomposition of the
+# positive-definite kernel matrix :math:`K``
```
Comment (rosecers) [inline examples/PlotPointwiseGFRE.py:53]: prime mark is misplaced

File: skmatter/datasets/descr/nice_dataset.rst [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 NICE dataset
 ############
+
```
Comment (rosecers) [inline skmatter/datasets/descr/nice_dataset.rst:5]: I don't understand why these changes are here...

File: docs/requirements.txt [hunk 5]
```diff
@@ -1,12 +1,6 @@
-ipykernel
 matplotlib
-nbsphinx==0.8.12
-nbconvert
-numpy
-scikit-learn >=0.24.0
+pandas
```
Comment (rosecers) [inline docs/requirements.txt:2]: Do the examples no longer require scikit-learn? I think it'd be good to have the explicit requirement.

**Review by rosecers — CHANGES_REQUESTED** [review]
Overall this looks nice, but I caught a few conversion artifacts here and there. I have two other issues that should be ameliorated before merge -- 

1. Ordering of the gallery -- we previously had the examples in a specific order, can we preserve this for the gallery?
2. Sidebar contains all sub-headers, and should not.

I've built the docs at https://scikit-matter.readthedocs.io/en/sphinx_gallery/index.html if you'd like to see.

**Comment by PicoCentauri** [comment]
Thanks for the feedback. I will see what I can do.

Manual ordering always works.

**Comment by PicoCentauri** [comment]
> I've built the docs at https://scikit-matter.readthedocs.io/en/sphinx_gallery/index.html if you'd like to see.

Thanks but you don't have to enable this manually the next time. In #161 we enabled that RTD builds the docs for each PR. For this one it is here: https://scikit-matter--170.org.readthedocs.build/en/170/. I will add a GH action in #172 for a direct link.

File: docs/requirements.txt (hunk 5 above)
Comment (PicoCentauri) [reply docs/requirements.txt:2]: They do but we have scikit-learn as an explicit dependency. I don't see a point of adding this here again.

File: examples/PCovR.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+#
+# Note on KernelPCovR for Atoms, Molecules, and Structures
+# --------------------------------------------------------
+#
+# Applying this to datasets involving collections of atoms and their atomic descriptors,
+# it's important to consider the nature of the property you are learning and the samples
+# you are comparing before constructing a kernel, for example, whether the analysis is
+# to be based on whole structures or individual atomic environments. For more detail,
+# see Appendix C of [Helfrecht 2020](https://iopscience.iop.org/article/10.1088/2632-2153/aba9ef)
+# or regarding kernels involving gradients [Musil 2021](https://arxiv.org/pdf/2101.08814.pdf).
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/PCovR.py:168]: Fixed.

File: examples/PlotGFRE.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    global_reconstruction_distortion,
+    global_reconstruction_error,
+)
+
+
+mpl.rc("font", size=20)
+
+# %%
+#
+# load features
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/PlotGFRE.py:40]: How should this be rendered? In the notebooks they were just comments in a code block. I converted them into text blocks.

File: examples/FeatureSelection-WHODataset.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    print(X_raw[:, columns.index(ls)].min(), X_raw[:, columns.index(ls)].max())
+    if ls in columns:
+        X_raw[:, columns.index(ls)] = np.log10(X_raw[:, columns.index(ls)])
+y_raw = np.array(df["SP.DYN.LE00.IN"])  # [np.where(df['Year']==2000)[0]])
+y_raw = y_raw.reshape(-1, 1)
+X_raw.shape
+
+# %%
+#
+# Scale and Center the Features and Targets
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/FeatureSelection-WHODataset.py:81]: Should be done in the latest version.

File: examples/OrthogonalRegressionNonAnalytic.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    orth_reg_xy_plane_pred_cube[:, 1],
+    c="g",
+    label="orth. reconstruction\n use projector=False",
+)
+
+ax_xy.set_title("xy plane")
+
+plt.legend(bbox_to_anchor=(1, 1), loc="upper left")
+
+plt.show()
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/OrthogonalRegressionNonAnalytic.py]: Should be done.

File: examples/PCovR_Scaling.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+ax2.set_xlabel(r"Mixing parameter $\alpha$")
+ax2.set_ylabel(r"Loss $\ell$")
+
+fig.show()
+
+# %%
+#
+# **Note: when the relative magnitude of the features or targets is important, such as
+# in load_csd_1000r, one should use the `StandardFlexibleScaler` provided by
+# ``scikit-matter``.**
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/PCovR_Scaling.py]: Should be done including links to the docs.

File: skmatter/datasets/descr/nice_dataset.rst (hunk 4 above)
Comment (PicoCentauri) [reply skmatter/datasets/descr/nice_dataset.rst]: They produced warnings in the doc building process. I removed them from this PR since I have to fix them in #172 anyway.

File: examples/OrthogonalRegressionNonAnalytic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rosecers) [reply examples/OrthogonalRegressionNonAnalytic.py]: Now it looks all messed up
image

File: examples/PlotGFRE.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rosecers) [reply examples/PlotGFRE.py:40]: Right now they are rendered as headers, but formatted like comments. It has to be one or the other...

File: examples/OrthogonalRegressionNonAnalytic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/OrthogonalRegressionNonAnalytic.py]: Ohh this looks nice 🤣.

File: examples/PlotGFRE.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/PlotGFRE.py:40]: They are rendered as normal text not as headers I think. But, I will revert this to comments at it was before.

File: examples/OrthogonalRegressionNonAnalytic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/OrthogonalRegressionNonAnalytic.py:242]: Lets see how this looks in the new commit.

File: examples/OrthogonalRegressionNonAnalytic.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    0,
+    regression_error_for_xy_plane_projected_cube_zero_padded,
+    label="Regression error for xy_plane_projected_cube",
+)
+ax_wo_orth.set_title(
+    "Orthogonal regression error for\n zero padded features\n (use_orthogonal_projector=False) ",
+)
+ax_wo_orth.set_xlabel("scaling in z direction")
+ax_wo_orth.legend(loc="upper right", bbox_to_anchor=(0.7, -0.2))
+plt.show()
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply examples/OrthogonalRegressionNonAnalytic.py]: Lets see how this looks in the new commit.

**Comment by victorprincipe** [comment]
Just went through the examples and made some typo/formatting fixes. Rest looks good.

**Comment by PicoCentauri** [comment]
Thanks @victorprincipe. All of them are super useful changes. I fixed another tiny formatting thing.

**Comment by PicoCentauri** [comment]
> I don't see the nice gallery on the auto-built readthedocs page at https://scikit-matter--170.org.readthedocs.build/en/170/tutorials.html#

I had to remove the gallery view in order to keep the same structure as it was before. Sorry, if we want to have the gallery back I can look later how we can order it. But I think it will be not super easy to maintain the section titles.

**Comment by ceriottm** [comment]
Bummer. The gallery was really pretty.

**Comment by PicoCentauri** [comment]
Yes, I was also a fan of it.

**Comment by PicoCentauri** [comment]
The only way I see is that we do subfolders for each current section and then we use [sub-gallery-order](https://sphinx-gallery.github.io/stable/configuration.html#sub-gallery-order).

**Comment by PicoCentauri** [comment]
Galleries are back. Let me know @rosecers and @ceriottm if this fine for you.

**Comment by ceriottm** [comment]
I like it a lot - it's also better to have separate pages, and I was finding the list of sub-headings of the notebooks rather confusing. IMO this is good to go.

**Review by rosecers — APPROVED** [review]
Looks good! Please squash before merging.

**Comment by PicoCentauri** [comment]
Thanks for your feedback! I can help you with rebasing your PR @ceriottm
